sympy/tensor/array/expressions/einsum_sympy.py

```diff
@@ -0,0 +1,349 @@
+    if len(contraction_indices) > 0:
+        array_contraction = ArrayContraction(base, *contraction_indices)
+        diagonalization_indices = ArrayContraction._push_indices_up(
+            array_contraction.contraction_indices, diagonalization_indices)
+    else:
+        array_contraction = base
+
+    # Wrap in an ArrayDiagonal object:
+    array_diagonal = (
+        ArrayDiagonal(array_contraction, *diagonalization_indices)
+        if len(diagonalization_indices) > 0 else
+        array_contraction
+    )
+
+    # Wrap in a PermuteDims object:
+    permu = (
+        PermuteDims(array_diagonal, index_order_old=order_after_diagonalization, index_order_new=path_dst)
+        if order_after_diagonalization != list(path_dst) else
+        array_diagonal
+    )
+    return permu
+
+
+def _parse_path_from_string(path):
+    path = str(path)
+    path = path.replace(" ", "").replace("\t", "").strip()
+    indices1, indices2 = path.split("->")
+    indices1_by_arg = indices1.split(",")
+    return indices1_by_arg, indices2
+
+
+def _check_path_compatibility_with_array_args(path_src, path_dst, array_args):
+    if len(path_src) != len(array_args):
+        raise ValueError("mismatch between indices and number of arrays")
+    for i, arg in zip(path_src, array_args):
+        if len(i) != get_rank(arg):
+            raise ValueError(f"wrong number of indices for {arg}")
+
+
+class Einsum(_ArrayExpr):
+    """
+    The Einsum class represents an operator equivalent to NumPy's
+    einsum. Unlike NumPy, it does not evaluate the expression, unless
+    .as_explicit() is called.
+
+    Examples
+    ========
+
+    >>> from sympy.tensor.array.expressions.einsum_sympy import Einsum
+    >>> from sympy.tensor.array.expressions import ArraySymbol
+    >>> A = ArraySymbol("A", (4, 4, 4))
+
+    >>> Einsum("abb->a", A)
+    Einsum(Str('abb->a'), A)
+
+    >>> Einsum("abb->a", A).as_array_expression()
+    ArrayContraction(A, (1, 2))
+    >>> Einsum("abc->bca", A).as_array_expression()
+    PermuteDims(A, (0 1 2))
+
+    >>> from sympy.abc import x, y, z, t
+    >>> from sympy import Array
+    >>> B = Array([[x, y], [z, t]])
+    >>> B
+    [[x, y], [z, t]]
+
+    >>> Einsum("aa->a", B).as_explicit()
+    [x, t]
+    >>> Einsum("aa->", B).as_explicit()
+    t + x
+    >>> Einsum("ab->ba", B).as_explicit()
+    [[x, z], [y, t]]
+    >>> Einsum("ab->a", B).as_explicit()
+    [x + y, t + z]
+    >>> Einsum("ab->b", B).as_explicit()
+    [x + z, t + y]
+    """
+
+    def __new__(cls, *args):
+        if len(args) == 0:
+            raise ValueError("invalid number of arguments")
+        path_str: Str
+        array_args = []
+        if isinstance(args[0], (str, Str)):
+            path_src, path_dst = _parse_path_from_string(str(args[0]))
+            if isinstance(args[0], str):
+                path_str = Str(args[0])
+            else:
+                path_str = args[0]
+            array_args.extend(args[1:])
```

Comment:
I think storing the array in the Einsum object is a mistake. I would separate them out. If necessary, there can be an AppliedEinsum object. The only real thing you need is the array shapes, but even then, it would be better to separate a pure einsum path from one applied to specific array shapes.